### ISSUE TITLE
Final on-chain parameters

### ISSUE THREAD

**Issue by pdyraga** [keep-network/keep-core#1545]
Finalize on-chain parameters.
- result publication block step
- group active time
- ticket submission timeout and group duration
- group profit fee
- DKG frequency

- curve shape (stake drop)
- when it starts getting effective / when it's off

**Comment by eth-r** [comment]
Stakedrop begins at June 8 (interval 1), interval 0 covers everything before June 8 so should be dated May 8. 6 scheduled intervals (0 to 5 defined in reward schedule; from 6 onwards every interval allocates 100%) where 0 has a small reward allocation, 1 has the largest, and later intervals' allocations decay at some rate that gives a pretty curve.

If we want the cost of getting a random number to be somewhere in the region of $10, 500k gwei per member/32M (0.032 ETH) per group would yield roughly $5.5 of group profit at current prices. If we assume a minimum of 3 blocks per entry (`b = 3`) we have roughly 1M entries a year at full utilization, which gives an easy calculation of maximum beacon revenue.

Result publication block step is currently also used for entry timeout, which at 3 gives 192 blocks or around 30 to 50 minutes for an entry to be submitted before the group times out. This seems reasonable, and we could add an alarm functionality that tries to alert a human operator e.g. 5~10 minutes into the timeout to possibly see what's the problem and avoid getting slashed. I wouldn't be comfortable going much lower before we get experience and data in production.

Ticket submission timeout and result publication blocks are basically adjusting the rate at which new DKGs/group selections can be performed. If DKG frequency is expressed as `1/r`, whether group creation is constrained by time or pooled money at maximum utilization depends on whether `b*r > ticketSubmissionTimeout + timeDKG + maybe(resultPublicationBlockStep * groupSize)` (money-constrained) or not (time-constrained). Time-constrained DKG means that at max utilization the beacon accumulates a large pool of DKG fee money, which it will then use afterwards to keep the number of groups high. This seems moderately preferable, especially if `r` is low (otherwise we can get a lot of groups).

With 39 blocks for ticket submission, and 72 blocks for DKG, and a result publication step of 3 blocks, the total amount of time group selection+DKG can take varies from 112 to 303 blocks. This means that `r =< 37` is guaranteed to be time-constrained with an entry every 3 blocks, and `r =< 100` can become time-constrained.

The `r` in DKG frequency has a necessary relationship with group lifespan: we want there to be at least `r` entries per lifespan to avoid running out of groups, unless we manually fund the DKG pool and/or create demand on entries. On the other hand, the number of groups is effectively capped by `groupLifespanInBlocks / max(b*r, maxDkgTime)`. 

The gas reimbursement for group creation is roughly 2M gas, and the reimbursement ceiling is 30 gwei. This means that DKG adds `120M / r` gwei to the entry fee. We also need to consider ticket submission costs, for which we can allocate maybe 200k gas per member, or 13M gas per group. At reimbursement ceiling this totals to roughly `400M / r` gwei per entry. With a group profit of `P` gwei, members will break even at `r >= 400M / P` when gas costs are at max; if `P = 32M` it implies that `r >= 13`.

If a block is created every 10 seconds, we get 8640 blocks per day. Because group selection + DKG takes at least 112 blocks with current parameters, the maximum number of groups that can be created in a day is 77. This gives the maximum size of the active group pool as `groupLifespanInDays * 77`. With an initial minimum stake of 100k, 77 groups gives 492.8M KEEP as the total collateral. Given that we aren't expecting all available KEEP to be staked straight away, this is likely to already be leveraged territory where the same stakes are backing multiple group members. The good news is that we don't need to care about the O(N) nature of skipping terminated groups, as the number of terminated groups can't ever go above 156.

If we think roughly 1000 active groups is a reasonable ceiling at max utilization, that suggests 14 days as the maximum lifespan. If we set `r = 30`, group creation becomes time-constrained at maximum utilization, and we'd need a bit over 2 entries each day to keep the beacon self-sufficient. At max gas cost and `P = 32M` roughly 1/3 of the incoming group profit would be spent on tickets.

TL;DR: here's what I'd go with unless there are requirements I'm not aware of:
- block step of `3` with `2` as the absolute minimum, ticket submission timeout of `11*3 + 6 = 39`, DKG block counts as the numbers we've been testing with
- DKG frequency of `1/20 ~ 1/30`
- group lifespan of `1~14` days
- group profit of `16M ~ 64M` gwei (`250k ~ 1M` gwei per member)
- stakedrop curve of whatever @mhluongo has drawn on the whiteboard, or if he doesn't have anything yet I can go draw a curve and measure it

**Comment by eth-r** [comment]
If we want to make DKG and group expiration faster, we can reduce the block step to `2` for the purpose of calculating ticket submission (`28` blocks) and DKG result submission (`<128` blocks), or even to `1` for the latter (`<64` blocks because timing out doesn't give punishments), and we could tighten up DKG synchronization if clients can handle it.

I would rather not touch entry timeouts at all, and definitely not below `128` blocks.

DKG frequency can go lower if we want, only constrained by either having that many entries within the group lifespan or being willing to subsidize manually. Higher than `1/10` would be expensive.

Group lifespans can go a bit lower if we speed up DKG, but I'd stay above `12h` to ensure that periods of low demand can be coped with. The ceiling of group lifespan is the undelegation period, and with O(1) signing group selection we don't incur massive costs from having many active groups so the question is whether the clients suffer performance losses if we do.

Group profit can be lower if we want entries to be inexpensive early on, as long as we don't push too hard on the DKG frequency.

**Comment by pdyraga** [comment]
Thanks for breaking this down @eth-r. It all makes sense to me. I'd also like to avoid going down below 3 blocks for publication steps. Even on testnet 3 blocks are sometimes not enough so we shouldn't be very aggressive here, especially early days.

Worth remembering entry price is currently the whole fee for opening ECDSA keep and it's also an anti-spam measure there so we shouldn't go too low with the price. I think what you proposed is very reasonable. 

I can work with what we got here and stakedrop curve can be calculated next week.

**Comment by pdyraga** [comment]
@eth-r What's the minimal stake initialization period you are comfortable with?

**Comment by pdyraga** [comment]
https://github.com/keep-network/keep-core/pull/1653 addresses all those parameters except DKG frequency. I created a separate issue for that: https://github.com/keep-network/keep-core/issues/1660
